platforms/allwinner-d1/core/src/drivers/sharp_display.rs

```diff
@@ -1,73 +1,393 @@
-        // * Every FRAME gets a 1 byte command.
-        //     * It is zero, unless we are toggling VCOM.
-        //     * VCOM must be toggled once per second...ish.
-        // * Foreach LINE (240x) - 52 bytes total:
-        //     * 1 byte for line number
-        //     * (400bits / 8 = 50bytes) of data (one bit per pixel)
-        //     * 1 "dummy" byte
-        // * At the END, we need a total of two dummy bytes, so one from the last line, + 1 more
-        //
-        // This is where the 52x240 + 1 + 1 buffer size comes from.
+        if color.luma() > 128 {
+            self.frame[y][byte_x] |= 1 << (bit_x as u8);
+        } else {
+            self.frame[y][byte_x] &= !(1 << (bit_x as u8))
+        }
+    }
+}
+
+impl Dimensions for FullFrame {
+    fn bounding_box(&self) -> embedded_graphics::primitives::Rectangle {
+        Rectangle::new(
+            Point { x: 0, y: 0 },
+            Size {
+                width: 400,
+                height: 240,
+            },
+        )
+    }
+}
+
+impl DrawTarget for FullFrame {
+    type Color = Gray8;
+
+    type Error = ();
+
+    fn draw_iter<I>(&mut self, pixels: I) -> Result<(), Self::Error>
+    where
+        I: IntoIterator<Item = Pixel<Self::Color>>,
+    {
+        for px in pixels {
+            self.set_px(px.0.x as usize, px.0.y as usize, px.1)
+        }
+        Ok(())
+    }
+}
+
+//////////////////////////////////////////////////////////////////////////////
+// Helper tasks
+//
+// Friends to help us make things happen
+//////////////////////////////////////////////////////////////////////////////
+
+/// VCom - Once a second, update the vcom flag which is sent in every message.
+///
+/// TODO(AJM): The Beepberry uses a GPIO to toggle this rather than a bit in the
+/// SPI messages. This struct should be expanded to toggle the EXT pin at this
+/// rate instead, possibly as an optional field.
+struct VCom {
+    kernel: &'static Kernel,
+    ctxt: Arc<Mutex<Context>>,
+}
+
+impl VCom {
+    pub async fn run(self) {
+        loop {
+            self.kernel.sleep(Duration::from_secs(1)).await;
+            let mut c = self.ctxt.lock().await;
+            c.vcom = !c.vcom;
+        }
+    }
+}
+
+/// Drawing task
+///
+/// This task draws whenever there are pending (or "dirty") changes to the frame
+/// buffer, or at a base rate of 2Hz. At the moment, we only do a full redraw to
+/// ensure that VCOM is updated periodically. In the future, we can likely send
+/// a "no-op" command instead of a full frame redraw if the dirty flag has not
+/// been set, as a power optimization or latency improvement.
+struct Draw {
+    kernel: &'static Kernel,
+    buf: FixedVec<u8>,
+    spim: SpiSenderClient,
+    ctxt: Arc<Mutex<Context>>,
+}
+
+impl Draw {
+    async fn run(mut self) {
+        loop {
+            let c = self.ctxt.lock().await;
+            // render into
+            let vc = 0x01 | if c.vcom { 0x02 } else { 0x00 };
+            self.buf.as_slice_mut()[0] = 0x01 | vc;
```

Comment:
wait, this gets or'd with `0x01` twice, is that supposed to happen?

Reply:
Ah, this looks like a half finished thought. I meant to move it all to 203, and then didn't remove the command bit from 204. Good catch (it would do nothing)